shortfin/python/shortfin_apps/llm/components/kvcache/trie_attention_cache.py

```diff
@@ -0,0 +1,400 @@
+        last_cached_node: TrieNode,
+        cached_pages: List[PageInfo],
+        newly_acquired_pages: List[PageInfo],
+    ):
+        self.cache = cache
+        self.tokens = tokens
+        self.last_cached_node = last_cached_node
+        self._pages = cached_pages + newly_acquired_pages
+        self.number_of_published_pages = len(cached_pages)
+        self._is_released = False
+
+    @property
+    def pages(self) -> List[PageInfo]:
+        return self._pages
+
+    def publish_pages_for_tokens(self, tokens, publish_incomplete_page=False) -> None:
+        """Make pages available in the cache for the specified tokens.
+
+        Args:
+            tokens_to_publish: Tokens to publish to the cache
+
+        Raises:
+            ValueError: If tokens don't match allocation or exceed available pages
+        """
+        # If we have more tokens, publish pages up to the incoming tokens.
+        # If incoming has more tokens, replace our tokens with incoming tokens and publish pages up to the incoming tokens.
+
+        def has_common_prefix(tokens1, tokens2):
+            for t1, t2 in zip(tokens1, tokens2):
+                if t1 != t2:
+                    return False
+            return True
+
+        if not has_common_prefix(self.tokens, tokens):
+            raise ValueError(
+                "Tokens provided in publish_pages do not match tokens in allocation"
+            )
+
+        if len(tokens) > len(self.tokens):
+            self.tokens = tokens
+
+        tokens_per_page = self.cache.tokens_per_page
+
+        number_of_pages_to_publish = len(tokens) / tokens_per_page
+        if publish_incomplete_page:
+            number_of_pages_to_publish = math.ceil(number_of_pages_to_publish)
+        else:
+            number_of_pages_to_publish = math.floor(number_of_pages_to_publish)
+
+        # Create token blocks for unpublished pages
+        start_token_index = self.number_of_published_pages * tokens_per_page
+        unpublished_tokens = [
+            tuple(self.tokens[i : i + tokens_per_page])
+            for i in range(start_token_index, len(self.tokens), tokens_per_page)
+        ]
+
+        unpublished_pages = self._pages[
+            self.number_of_published_pages : number_of_pages_to_publish
+        ]
+
+        # Add unpublished pages to trie
+        if publish_incomplete_page:
+            raise NotImplementedError(
+                "Additional work needed here to support publishing incomplete pages to ensure that we finish up a page before attaching child nodes to it."
+            )
+        cur_node = self.last_cached_node
+        for token_block, page in zip(unpublished_tokens, unpublished_pages):
+            new_node = cur_node.create_child(token_block, page)
+            cur_node = new_node
+
+        if cur_node is not self.cache.root:
+            self.cache.leaves.add(cur_node)
+
+        # Update reference counts
+        if unpublished_tokens:
+            cur_node.ref_count += 1
```

Comment:
I'm wondering if we should be incrementing the `ref_count` for each node in lines 157-159. 

It makes sense to only update the ref count of the tail, since we would be evicting nodes based on their `ref_count` in a recursive DFS fashion.

However, I don't have a specific example on hand, but I'm wondering if we could run into confusion.

For example, let's say we published 3 pages for a given input of tokens. Then, nodes A -> B -> C would be referring to the same token input, just paged across 3 pages & nodes. 

But, only C would have a reference count incremented, even though A and B would still be referring to a subset of tokens that are part of the same input as C. 

So we end up with:

parent_node (ref_count = n) -> A (ref_count = 0) -> B (ref_count = 0) -> C (ref_count = 1)

Although, that could cause some issues, because we would need to update the entire chain of parents if we were to decrement or increment the tail, and there might not really be a way to know if a parent was part of the same token inputs, or was just already there.

Let me know what you think.

Reply:
This ref count works a little differently from what we usually use ref counts for - -

Usually we make something elegible for eviction when its ref count reaches 0.

Here, a node has to:
- be a leaf node
- have ref_count == 0
for eviction eligibility.

Here's a way to visualize the current eviction algo and how it interacts with ref counts.

when we need to evict some pages, we go through the leaves and pluck them one by one. whenever we make a node into a new leaf by plucking all of its children, we make it eligible for plucking also.

However, leaves with ref count > 0 are not eligible for plucking

which means that any nodes that have them as children will also not be pluckable.

Reply:
Gotcha, that makes sense. So, IIUC, having the parents' ref_count set to `0` makes that algorithm work. We pluck the first child (C), then can pluck B, A due to the fact that their ref counts are `0` and they no longer have children.

Reply:
tl;dr:

the TRUE reference count of any given node is actually

node.ref_count + the TRUE reference count of all its children.

this serves to make it easier to update reference counts without having to perform tree traversals.